Start moving "common" functionality to "transient-extras"

swflint opened the pull request.

swflint committed:
Start moving "common" functionality to "transient-extras"

swflint committed:
Move rotating switches to transient-extras

haji-ali commented on transient-extras.el:
```diff
@@ -0,0 +1,115 @@
+                   (apply-partially #'format (oref obj argument-format))
+                   (mapcar
+                    (lambda (x)
+                                        ; Return car of X if it is a cons, otherwise return X.
```
Extra spacing.

haji-ali commented on transient-extras.el:
```diff
@@ -0,0 +1,115 @@
+              choices
+              (propertize "|" 'face 'transient-inactive-value))))))
+
+;; TODO: Maybe use `choices' slot to `transient-init-value'
```
I think `transient-init-value` works as expected for `transient-extras-exclusive-switch`.
`transient-extras-exclusive-switch` inherits from `transient-switches` (the latter is also an exclusive switch).
The only difference is that `transient-extras-exclusive-switch` allows for a different label to be displayed for options values (for example displaying `draft/normal/best` instead of `3/4/5`).

PS: I am actually planning to propose this change upstream at some point.

swflint replied:
Does it do it without requiring specification of a regular expression?

haji-ali replied:
Yes, same as `transient-switches`.

swflint replied:
Cool!

haji-ali commented on transient-extras.el:
```diff
@@ -0,0 +1,115 @@
+
+;;; Files Lists
+
+;; TODO: (Planned feature) A way to add/remove files instead of replacing.
+;; For example by pressing:
+;; `-+': add files
+;; `-=': remove files
+;; `--': replace files
```
This probably should be removed since I copied it to as Github issue.

swflint replied:
Will do

haji-ali commented:
Looks great! Thank you.

Thinking more about it. I should probably rename the package to `transient-extras` and the file names would be `transient-extras-common` and  `transient-extras-lp` (and later  `transient-extras-a2ps` and `transient-extras-find`).

swflint commented:
I'm good with that naming convention (and I can make that change shortly).

I'm currently working on reimplementing the logic for reading the printer as something that's a bit more generic as well.

swflint committed:
Fix spacing issue

swflint committed:
Drop TODO comment

swflint committed:
Make logic to read printer names more generic

This allows other transient-extras to read options from the command
line using the same logic and reducing rewritten code.

swflint committed:
Rename transient-extras to transient-extras-common

swflint committed:
Add an author line

swflint committed:
Add initforms to simplify boundedness checking

swflint committed:
Boundedness checking is no longer necessary

swflint committed:
Add & clean up some commentary

swflint committed:
Add a TODO

haji-ali commented:
Thanks @swflint, this looks about ready to merge. There are some long lines and formatting issues, but I can deal with those myself before merging (you can see those with `checkdoc` or `package-lint`).

More importantly, the change to the printer reading is a significant change (at least for my setup) rather than a change in naming/package organization. I am keen to merge the changes related to naming/organization and discuss the change in printer-name in a separate pull-request.

swflint commented:
I'll re-write the PR this evening to separate printer reading and have `checkdoc` run.

swflint commented:
Given #7 and #8, this is closed.
